File: script/src/sensors.c
// Native bindings: the sensors + location half of "zelto:native" (P38).
//
// Same shape as sys.c — thin C that unmarshals JS args, calls the libzelto
// z_sensor_* / z_loc_* client, and marshals results back. Streams (a sensor or a
// GPS watch) park a long-lived JS callback in the registry and fire it per sample
// from the app loop; a one-shot location read is synchronous. Rate presets and
// the value shape mirror what the JS wrapper (zelto/sensors) and the andemu
// bridge (android/hardware) expect.
#include <stdint.h>
#include <string.h>

#include "internal.h"
#include "zelto/ui.h"

// Wire names indexed by ZSensorType (tracks the enum in <zelto/ui.h>).
static const char *const zs_sensor_names[Z_SENSOR_COUNT] = {
    "accelerometer", "gyroscope",  "magnetometer", "orientation",
    "gravity",       "linear_acceleration", "rotation_vector", "light",
    "proximity",     "pressure",   "step_counter",
};

static bool require_app(JSContext *ctx) {
    if (!zs_current()->app) {
        JS_ThrowTypeError(ctx, "the app is not running yet (call this from a "
                               "component, an effect, or a handler)");
        return false;
    }
    return true;
}

static int sensor_type_of(const char *name) {
    for (int i = 0; i < Z_SENSOR_COUNT; i++) {
        if (strcmp(zs_sensor_names[i], name) == 0) {
            return i;
        }
    }
    return -1;
}

// handle -> parked callback index, so sensorClose/locationStop can release the
// registry slot when a stream ends (a polling app reuses one slot).
#define ZS_SENSOR_MAX 64
static struct {
    int handle;
    uint32_t cbidx;
} zs_sensor_cbs[ZS_SENSOR_MAX];

static void sensor_cb_track(int handle, uint32_t cbidx) {
    for (int i = 0; i < ZS_SENSOR_MAX; i++) {
        if (zs_sensor_cbs[i].handle == 0) {
            zs_sensor_cbs[i].handle = handle;
            zs_sensor_cbs[i].cbidx = cbidx;
            return;
        }
    }
}

// Release the callback slot for a handle (returns true if one was tracked).
static bool sensor_cb_release(int handle) {
    for (int i = 0; i < ZS_SENSOR_MAX; i++) {
        if (zs_sensor_cbs[i].handle == handle) {
            zs_cb_del(zs_current()->ctx, zs_sensor_cbs[i].cbidx);
            zs_sensor_cbs[i].handle = 0;
            return true;
        }
    }
    return false;
}

// --- sensors ---------------------------------------------------------------

static void sensor_sample_cb(ZApp *app, const ZSensorSample *s, void *ud) {
    (void)app;
    JSContext *ctx = zs_current()->ctx;
    JSValue fn = zs_cb_get(ctx, (uint32_t)(uintptr_t)ud);

    JSValue ev = JS_NewObject(ctx);
    JS_SetPropertyStr(ctx, ev, "type",
                      JS_NewString(ctx, (s->type >= 0 && s->type < Z_SENSOR_COUNT)
                                            ? zs_sensor_names[s->type]
                                            : "unknown"));
    JSValue arr = JS_NewArray(ctx);
    for (int i = 0; i < s->n && i < 3; i++) {
        JS_SetPropertyUint32(ctx, arr, (uint32_t)i, JS_NewFloat64(ctx, s->v[i]));
    }
    JS_SetPropertyStr(ctx, ev, "values", arr);
    JS_SetPropertyStr(ctx, ev, "accuracy", JS_NewInt32(ctx, s->accuracy));
    JS_SetPropertyStr(ctx, ev, "t", JS_NewInt64(ctx, s->t));

    zs_call_handler(fn, 1, (JSValueConst *)&ev);
    JS_FreeValue(ctx, ev);
    JS_FreeValue(ctx, fn);
}

// sensorOpen(typeName, rateHz, fn) -> handle (0 on failure / unknown sensor)
static JSValue js_sensor_open(JSContext *ctx, JSValueConst this_val, int argc,
                              JSValueConst *argv) {
    (void)this_val;
    if (argc < 3 || !JS_IsFunction(ctx, argv[2])) {
        return JS_ThrowTypeError(ctx, "sensorOpen(type, rateHz, fn)");
    }
    if (!require_app(ctx)) { return JS_EXCEPTION; }
    const char *name = JS_ToCString(ctx, argv[0]);
    if (!name) { return JS_EXCEPTION; }
    int type = sensor_type_of(name);
    JS_FreeCString(ctx, name);
    if (type < 0) { return JS_NewInt32(ctx, 0); }   // unknown sensor: not present
    int32_t rate = Z_SENSOR_RATE_NORMAL;
    if (argc > 1) { JS_ToInt32(ctx, &rate, argv[1]); }

    uint32_t cbidx = zs_cb_add(ctx, argv[2]);
    int handle = z_sensor_open(zs_current()->app, (ZSensorType)type, rate,
                               sensor_sample_cb, (void *)(uintptr_t)cbidx);
    if (handle == 0) {
        zs_cb_del(ctx, cbidx);
        return JS_NewInt32(ctx, 0);
    }
    sensor_cb_track(handle, cbidx);
    return JS_NewInt32(ctx, handle);
}

static JSValue js_sensor_set_rate(JSContext *ctx, JSValueConst this_val, int argc,
                                  JSValueConst *argv) {
    (void)this_val;
    if (argc < 2) { return JS_UNDEFINED; }
    int32_t handle = 0, rate = 0;
    JS_ToInt32(ctx, &handle, argv[0]);
    JS_ToInt32(ctx, &rate, argv[1]);
    z_sensor_set_rate(handle, rate);
    return JS_UNDEFINED;
}

static JSValue js_sensor_close(JSContext *ctx, JSValueConst this_val, int argc,
                               JSValueConst *argv) {
    (void)this_val;
    if (argc < 1) { return JS_UNDEFINED; }
    int32_t handle = 0;
    JS_ToInt32(ctx, &handle, argv[0]);
    z_sensor_close(handle);
    sensor_cb_release(handle);
    return JS_UNDEFINED;
}

// sensorPresent(typeName) -> boolean
static JSValue js_sensor_present(JSContext *ctx, JSValueConst this_val, int argc,
                                 JSValueConst *argv) {
    (void)this_val;
    if (argc < 1) { return JS_FALSE; }
    const char *name = JS_ToCString(ctx, argv[0]);
    if (!name) { return JS_EXCEPTION; }
    int type = sensor_type_of(name);
    JS_FreeCString(ctx, name);
    if (type < 0) { return JS_FALSE; }
    ZSensorCaps caps = z_sensor_info((ZSensorType)type);
    return JS_NewBool(ctx, caps.present);
}

// --- location --------------------------------------------------------------

static JSValue location_to_js(JSContext *ctx, const ZLocation *loc) {
    if (!loc->ok) { return JS_NULL; }
    JSValue o = JS_NewObject(ctx);
    JS_SetPropertyStr(ctx, o, "lat", JS_NewFloat64(ctx, loc->lat));
    JS_SetPropertyStr(ctx, o, "lng", JS_NewFloat64(ctx, loc->lng));
    JS_SetPropertyStr(ctx, o, "accuracy", JS_NewFloat64(ctx, loc->accuracy));
    JS_SetPropertyStr(ctx, o, "altitude", JS_NewFloat64(ctx, loc->altitude));
    JS_SetPropertyStr(ctx, o, "speed", JS_NewFloat64(ctx, loc->speed));
    JS_SetPropertyStr(ctx, o, "bearing", JS_NewFloat64(ctx, loc->bearing));
    JS_SetPropertyStr(ctx, o, "t", JS_NewInt64(ctx, loc->t));
    return o;
}

// locationGet() -> { lat, lng, ... } | null   (synchronous fast read)
static JSValue js_location_get(JSContext *ctx, JSValueConst this_val, int argc,
                               JSValueConst *argv) {
    (void)this_val; (void)argc; (void)argv;
    if (!require_app(ctx)) { return JS_EXCEPTION; }
    ZLocation loc = z_loc_get(zs_current()->app);
    return location_to_js(ctx, &loc);
}

static void location_watch_cb(ZApp *app, const ZLocation *loc, void *ud) {
    (void)app;
    JSContext *ctx = zs_current()->ctx;
    JSValue fn = zs_cb_get(ctx, (uint32_t)(uintptr_t)ud);
    JSValue ev = location_to_js(ctx, loc);
    zs_call_handler(fn, 1, (JSValueConst *)&ev);
    JS_FreeValue(ctx, ev);
    JS_FreeValue(ctx, fn);
}

// locationWatch(rateHz, fn) -> handle (0 on failure)
static JSValue js_location_watch(JSContext *ctx, JSValueConst this_val, int argc,
                                 JSValueConst *argv) {
    (void)this_val;
    if (argc < 2 || !JS_IsFunction(ctx, argv[1])) {
        return JS_ThrowTypeError(ctx, "locationWatch(rateHz, fn)");
    }
    if (!require_app(ctx)) { return JS_EXCEPTION; }
    int32_t rate = Z_SENSOR_RATE_MIN;
    JS_ToInt32(ctx, &rate, argv[0]);

    uint32_t cbidx = zs_cb_add(ctx, argv[1]);
    int handle = z_loc_watch(zs_current()->app, rate, location_watch_cb,
                             (void *)(uintptr_t)cbidx);
    if (handle == 0) {
        zs_cb_del(ctx, cbidx);
        return JS_NewInt32(ctx, 0);
    }
    sensor_cb_track(handle, cbidx);
    return JS_NewInt32(ctx, handle);
}

static JSValue js_location_stop(JSContext *ctx, JSValueConst this_val, int argc,
                                JSValueConst *argv) {
    (void)this_val;
    if (argc < 1) { return JS_UNDEFINED; }
    int32_t handle = 0;
    JS_ToInt32(ctx, &handle, argv[0]);
    z_loc_stop(handle);
    sensor_cb_release(handle);
    return JS_UNDEFINED;
}

// --- module table ----------------------------------------------------------

const JSCFunctionListEntry zs_sensor_funcs[] = {
    JS_CFUNC_DEF("sensorOpen", 3, js_sensor_open),
    JS_CFUNC_DEF("sensorSetRate", 2, js_sensor_set_rate),
    JS_CFUNC_DEF("sensorClose", 1, js_sensor_close),
    JS_CFUNC_DEF("sensorPresent", 1, js_sensor_present),
    JS_CFUNC_DEF("locationGet", 0, js_location_get),
    JS_CFUNC_DEF("locationWatch", 2, js_location_watch),
    JS_CFUNC_DEF("locationStop", 1, js_location_stop),
};

const int zs_sensor_funcs_count =
    (int)(sizeof(zs_sensor_funcs) / sizeof(zs_sensor_funcs[0]));

File: script/js/sensors.js
// The `zelto/sensors` module: device sensors + location, shaped as the docs
// promise (docs/system-apis/location.md).
//
// A stream (location.watch, sensors.gyroscope, ...) is driven by the zsysd sensor
// source over libzelto's async socket, parked in the app loop — samples arrive
// between frames, nothing blocks the UI. Location needs the `location` permission
// and the other sensors need `sensors`; a stream awaits its grant first, so a
// denial rejects rather than silently doing nothing (the zelto/net idiom).
//
// Refresh rate is Hz, clamped by the platform to [1, 60] (60 is display-aligned;
// higher is wasted on a software-rendered device). The named presets mirror the
// Android SensorManager delays, which is what the andemu bridge maps onto.
import * as N from "zelto:native";
import { ensure, status } from "zelto/permissions";

// Subscribe once the grant is held. The broker only allows ONE consent request in
// flight at a time, so a "request once, open many" app (the common shape — open
// the accelerometer AND the magnetometer to fuse orientation) must not fire a
// fresh consent per stream. When the capability is already granted we subscribe
// straight away; only an undecided/denied grant awaits the dialog. That mirrors a
// native app, which calls z_perm_request once and then opens every stream.
function whenGranted(perm, start) {
  if (status(perm) === "granted") {
    start();
    return;
  }
  ensure(perm).then(start).catch(() => {});
}

export const Rate = {
  min: 1,
  normal: 5,
  ui: 16,
  game: 50,
  fastest: 60,
  max: 60,
};

// --- location --------------------------------------------------------------

export const location = {
  // "granted" | "denied" | "prompt"
  permission() {
    const s = N.permStatus("location");
    if (s === N.PERM_GRANTED) return "granted";
    if (s === N.PERM_DENIED) return "denied";
    return "prompt";
  },

  // A single fix: { lat, lng, accuracy, altitude, speed, bearing, t } (null if
  // denied / no fix). Awaits the grant first.
  async current() {
    await ensure("location");
    return N.locationGet();
  },

  // Synchronous last fix — no await, no prompt. Returns the fix if the grant is
  // already held, else null (the broker gates it). This is what Android's
  // synchronous getLastKnownLocation maps onto.
  lastKnown() {
    return N.locationGet();
  },

  // Continuous updates. `opts.rate` is Hz (default 1). Returns an unsubscribe fn.
  // A denied grant throws (await it) rather than starting a dead stream.
  watch(cb, opts = {}) {
    const rate = opts.rate ?? Rate.min;
    let handle = 0;
    let stopped = false;
    whenGranted("location", () => {
      if (stopped) return;
      handle = N.locationWatch(rate, (p) => { if (p) cb(p); });
    });
    return () => {
      stopped = true;
      if (handle) N.locationStop(handle);
    };
  },
};

// --- sensors ---------------------------------------------------------------

// Whether the device exposes a sensor type ("gyroscope", "accelerometer", ...).
export function has(type) {
  return N.sensorPresent(type);
}

// Open a sensor stream. `cb` receives { type, values:[...], accuracy, t } per
// sample. `opts.rate` is Hz (default 5). Returns an unsubscribe fn. Awaits the
// `sensors` grant; a denial simply never delivers.
export function open(type, cb, opts = {}) {
  const rate = opts.rate ?? Rate.normal;
  let handle = 0;
  let stopped = false;
  whenGranted("sensors", () => {
    if (stopped) return;
    handle = N.sensorOpen(type, rate, (ev) => cb(ev));
  });
  return () => {
    stopped = true;
    if (handle) N.sensorClose(handle);
  };
}

// Named helpers for the common sensors — each is `open` bound to a type.
export const accelerometer = (cb, opts) => open("accelerometer", cb, opts);
export const gyroscope = (cb, opts) => open("gyroscope", cb, opts);
export const magnetometer = (cb, opts) => open("magnetometer", cb, opts);
export const orientation = (cb, opts) => open("orientation", cb, opts);
export const gravity = (cb, opts) => open("gravity", cb, opts);
export const light = (cb, opts) => open("light", cb, opts);
export const proximity = (cb, opts) => open("proximity", cb, opts);
export const pressure = (cb, opts) => open("pressure", cb, opts);
export const stepCounter = (cb, opts) => open("step_counter", cb, opts);
